File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        currMax = 0

        i = 0
        j = 1

        while j < len(prices):
            if prices[i] < prices[j]: #profit
                nMax =  prices[j] - prices[i]
                currMax = max(currMax, nMax)
                j += 1
            else:
                #move i to next one
                i = j
                j += 1
        
        return currMax

 

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() <= 1) {
            return 0;
        }

        int minPrice = prices[0];

        int best = 0;

        for (int i = 0 ; i < prices.size(); i++) {
            int prof = prices[i] - minPrice;

            if (prof > best) {
                best = prof;
            }

            if (prices[i] < minPrice) {
                minPrice = prices[i];
            }

        }
        return best;


    }
};